### PR TITLE
Add Gleam implementation

### PR DESCRIPTION
**Pull request by chiroptical**
# READ ME BEFORE SUBMITTING A PR

Please do not submit a PR with your solution to the Gilded Rose Kata. This repo is intended to be used as a starting point for the kata.

- [x] I acknowledge that this PR is not a solution to the Gilded Rose Kata, but an improvement to the template.

## Please provide your PR description below this line

Hello, I would like to add a starting point for the gleam programming language https://gleam.run/

I was able to confirm that the implementation passes the text tests

```
➤ gleam run > ~/output.txt
➤ diff ../texttests/ThirtyDays/stdout.gr ~/output.txt
0a1,2
>    Compiled in 0.02s
>     Running gilded_rose.main
373d374
< 
```

I didn't add this to the text test configuration because gleam doesn't produce an executable. I am not familiar enough with the tool to know if it would support something like `gleam run`. Also, `gleam run` does produce a bit of compiler information into stdout so it might cause test text to fail.

### CODE REVIEW

File: gleam/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+## Get the TextTest fixture on the command line
+
+Execute it on the command line with an argument for the number of days:
+
+```sh
+$ gleam run -- --days=30
+```
+
+Compare the output to [stdout.gr](../texttests/ThirtyDays/stdout.gr) for validation.
+There are instructions in the [TextTest Readme](../texttests/README.md) for setting up TextTest.
```
Comment (chiroptical) [inline gleam/README.md:20]: I could instead replace this with the

```
$ gleam run > /tmp/output.txt
$ diff /tmp/output.txt ../texttests/ThirtyDays/stdout.gr
```

If preferred.

File: gleam/manifest.toml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# This file was generated by Gleam
```
Comment (chiroptical) [inline gleam/manifest.toml:1]: I guess I should remove this!

File: gleam/src/gilded_rose.gleam [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  let days_flag =
+    flag.int()
+    |> flag.default(30)
+    |> flag.description("Number of days")
+
+  let flag = "days"
+
+  let simulate_inventory = fn() {
+    use input <- glint.command()
+    let assert Ok(number_of_days) = flag.get_int(from: input.flags, for: flag)
```
Comment (chiroptical) [inline gleam/src/gilded_rose.gleam:140]: Note: I noticed other examples provided an option to select a different number of days. I don't exactly know why and this does bring in more dependencies. If `gleam run` should only produce 30 days of output I can strip that all out.

Comment (chiroptical) [reply gleam/src/gilded_rose.gleam:140]: It would drop both `glint` and `argv`

File: gleam/README.md (hunk 1 above)
Comment (codecop) [reply gleam/README.md:20]: It's ok as provi

File: gleam/src/gilded_rose.gleam (hunk 3 above)
Comment (codecop) [reply gleam/src/gilded_rose.gleam:140]: Option for days is required but default is not 30 but 2 in most languages. Please change that. Also can we split the main code to file named "Program" or "TextTestFixture" please

**Review by codecop — CHANGES_REQUESTED** [review]
Looks very good otherwise. Thank you.

File: gleam/src/gilded_rose.gleam (hunk 3 above)
Comment (chiroptical) [reply gleam/src/gilded_rose.gleam]: I made an assumption that you preferred squashed commits. Hopefully that was accurate. I set the default to 2 here: https://github.com/emilybache/GildedRose-Refactoring-Kata/pull/548/commits/c9f452655476985f795a7d9a942bc2ecd32d7f70#diff-506be539568729aa4eb6af30a87b856b47ac987a0b7e48f7e428faf15ab81339R13

The structure is now

-  `src/cli.gleam` which contains the cli application
- `src/gilded_rose.gleam` contains my main function that runs the cli app
- `src/gilded_rose_item.gleam` contains the gilded rose item and update function
- `src/test_fixture.gleam` contains the test fixture and simulation function

Comment (chiroptical) [reply gleam/src/gilded_rose.gleam]: Thanks for the review!

File: gleam/test/gilded_rose_test.gleam
Comment (codecop) [inline gleam/test/gilded_rose_test.gleam:1]: This file is ok

**Comment by codecop** [comment]
Sorry for requesting another change I did not communicate clearly. It is very important to have as much the same file structure as other languages to avoid confusion of participants which could work in different languages. I have indicated what is needed

**Comment by chiroptical** [comment]
> Sorry for requesting another change I did not communicate clearly. It is very important to have as much the same file structure as other languages to avoid confusion of participants which could work in different languages. I have indicated what is needed

Sorry, I looked at a few other functional programming languages and I don't really see a similar structure. Could you clarify what you want where? Or provide a language that exhibits this structure and I'll copy it.

> main code to file named "Program" or "TextTestFixture"

Does this mean `mv src/gilded_rose_item.gleam src/program.gleam`? The literal `main` function needs to exist in `src/gilded_rose.gleam` but I am happy to move code wherever you prefer.

**Comment by chiroptical** [comment]
It is also possible you mean that `update_quality` **must** exist in `src/gilded_rose.gleam` but the cli/simulation code can be elsewhere?

**Comment by chiroptical** [comment]
https://github.com/emilybache/GildedRose-Refactoring-Kata/pull/548/commits/9f9971ebac22d10ff8fdf273cc22018777e7a5da

I think you mean this.

**Comment by codecop** [comment]
Other languages usually have 3 files 
* Gilded_rose with types. Item. Update quality 
* TextTestFixture or Program or main with all simulation code. Including dynamic days
* Gilded_rose-test or similar name with initial test depending on language conventions. The test you have is fine 
Is this possible to create with gleam infrastructure? You mention that gilded_rose must contain the main method? By language design? Is it because the module is named like that? Let's discuss before you modify code to avoid misunderstanding

**Comment by chiroptical** [comment]
> You mention that gilded_rose must contain the main method?

https://github.com/emilybache/GildedRose-Refactoring-Kata/pull/548/files#diff-9cddb0c6718f48079ff3119fb817294d5d7a17c8e08984f35ecb2c0307067434R1 <- this line dictates where the `main` function has to be. I could, for example, modify this to `program` and then `program.gleam` would contain the `main` function. That would allow me to move the `Item` back into `gilded_rose` (it had to be separate due to circular dependencies). I could easily put the test fixture, simulation, and cli code into `program.gleam` it just didn't feel like good naming. However, I understand you are looking for more uniformity across the project.

My proposal,

1. Change `gleam.toml` `name` to `program` 
2. Move `Item` and `GildedRose` back into `gilded_rose.gleam`
3. Move the simulation, cli app, and main function into `program.gleam`
4. Leave the test code alone.

Going to quickly confirm this is all possible. I'll add another comment when I confirm.

Thanks so much for working with me here.

**Comment by chiroptical** [comment]
All of that proposal works, only one change. In the test code, I would need to move the "test main" function into `test/program_test.gleam`. The `test/gilded_rose_item.gleam` can stay. The error from `gleam test` looks like

```command
> gleam test
   Compiled in 0.01s
    Running program_test.main
F
Failures:

  1) gilded_rose_test.update_quality_test: module 'gilded_rose_test'
     Values were not equal
     expected: "fixme"
          got: "foo"
     output:

Finished in 0.005 seconds
1 tests, 1 failures
```

so it should be clear where the test resides.

**Comment by codecop** [comment]
Yes exactly. Just confirm, the file name will be gilded_rose.gleam and not gilded_rose_item.gleam. Yes please make the changes.

**Comment by chiroptical** [comment]
> Yes exactly. Just confirm, the file name will be gilded_rose.gleam and not gilded_rose_item.gleam. Yes please make the changes.

Yup! It will contain the `Item` and `GildedRose` types and the `update_quality` function.

**Comment by chiroptical** [comment]
In theory, all set. Let me know if you want me to squash and force push or anything. Thanks for working with me on this.

Any interest in me writing a `CONTRIBUTING.md` quick (in another PR)? Might help shore up some confusion when adding languages?

**Comment by codecop** [comment]
Thank you. Yes if you can provide a CONTRIBUTING.md, please have a separate PR for that.

**Pull request merged**